src/Microsoft.AspNet.OData.Shared/Query/ODataPathQueryBuilder.cs

```diff
@@ -0,0 +1,309 @@
+        {
+            ParameterExpression param = Expression.Parameter(queryable.ElementType);
+            MemberExpression propertyExpression = Expression.Property(param, propertyName);
+
+            // for collection properties we use SelectMany instead of Select
+            // because Select would return an IQueryable<IEnumerable<TResult>>,
+            // but SelectMany flattens it to IQueryable<TResult>
+
+            // we don't check whether the property is null before advancing
+            // because that seems to cause exceptions in EF
+
+            // We expect the collection navigation property to implements IEnumerable<T>
+            // Here we extract the element type T
+            Type collectionPropertyElementType = GetEnumerableItemType(propertyExpression.Type);
+
+            // The lambda passed to queryable.SelectMany() has a signature Func<TSource, IEnumerable<TResult>>
+            // TSource corresponds to our queryable.ElementType and TResult is the element type of the nav property
+            Type delegateType = typeof(Func<,>).MakeGenericType(
+                queryable.ElementType,
+                typeof(IEnumerable<>).MakeGenericType(collectionPropertyElementType));
+            LambdaExpression selectBody =
+                Expression.Lambda(delegateType, propertyExpression, param);
+
+            // return queryable.SelectMany(entity => entity.CollectionNavProp)
+            return ExpressionHelpers
+                .SelectMany(queryable, selectBody, queryable.ElementType, collectionPropertyElementType);
+        }
+
+        /// <summary>
+        /// Gets the element type "T" of the given type
+        /// if it implements <see cref="IEnumerable{T}"/>
+        /// </summary>
+        /// <param name="enumerableType"></param>
+        /// <returns>
+        /// The element type if <paramref name="enumerableType"/> implements <see cref="IEnumerable{T}"/>
+        /// otherwise returns <paramref name="enumerableType"/> itself
+        /// </returns>
+        private static Type GetEnumerableItemType(Type enumerableType)
+        {
+            Type type = FindGenericType(enumerableType, typeof(IEnumerable<>));
+            if (type != null)
+            {
+                return type.GetGenericArguments()[0];
+            }
+
+            return enumerableType;
+        }
+
+        /// <summary>
+        /// Find a base type or implemented interface which has a generic definition
+        /// represented by the parameter, <c>definition</c>.
+        /// </summary>
+        /// <param name="type">
+        /// The subject type.
+        /// </param>
+        /// <param name="definition">
+        /// The generic definition to check with.
+        /// </param>
+        /// <returns>
+        /// The base type or the interface found; otherwise, <c>null</c>.
+        /// </returns>
+        private static Type FindGenericType(Type type, Type definition)
```

Comment:
A better home for this method would be a static TypeHelpers class - one might exist

Reply:
I removed this by reusing methods in `TypeHelper`, thanks.